File: Programming_cheat_sheets/_scripting_languages/Javascript/_data_structures/hash-table/hashTable.js
// hash table

function Entry(key, value){
    var self = this;
    self.key = key;
    self.value = value;
    self.next = null;
}

function HashTable(size){
    var self = this;
    
    //
    // public attributes
    //
    
    self.size = size;
    self.table = [];
    
    //
    // public functions
    //
    
    self.inset = inset;
    self.get = get;
    
    //
    // initialization
    //

    create(self.size);
    
    //
    // private functions
    //

    function create(){
        for(var i = 0; i < self.size; i++){
            self.table[i] = null;
        }
    }

    function strcmp(a, b){   
        return ( a < b ? -1 : ( a > b ? 1 : 0 ) );  
    }
    
    /**
     *
     */
    function hashFunction(key){
        var hashval = 0,
            i = 0,
            ULONG_MAX = 4294967295;

        /* Convert our string to an integer */
        while(hashval < ULONG_MAX && i < key.length){
            hashval = hashval << 8; // like multiplying x by (2 to the power of 8  (256))
            asciiVal = key[i].charCodeAt(0);
            hashval += asciiVal;
            i++;
        }

        return hashval % self.size;
    }
    
    /**
     *
     */
    function createEntry(key, value){
        return new Entry(key, value);
    }
    
    /**
     *
     */
    function inset(key, value){
        var index = hashFunction(key);

        next = self.table[index];

        while(next != null && next.key != null && strcmp(key, next.key) > 0){
            last = next;
            next = next.next;
        }

        /* There's already an entry.  Let's replace that string. */
        if(next != null && next.key != null && strcmp(key, next.key) == 0){
            next.value = null;
            next.value = value;

        /* Nope, could't find it.  Time to grow an entry. */
        }else{
            newEntry = createEntry(key, value);

            /* We're at the start of the linked list in this index. */
            if(next == self.table[index]){
                newEntry.next = next;
                self.table[index] = newEntry;

            /* We're at the end of the linked list in this index. */
            }else if(next == null){
                last.next = newEntry;

            /* We're in the middle of the list. */
            }else{
                newEntry.next = next;
                last.next = newEntry;
            }
        }
    }
    
    /**
     *
     */
    function get(key){
        var index = hashFunction(key);

        /* Step through the index, looking for our value. */
        var entry = self.table[index];
        while(entry != null && entry.key != null && strcmp(key, entry.key) > 0){
            entry = entry.next;
        }

        /* Did we actually find anything? */
        if(entry == null || entry.key == null || strcmp(key, entry.key) != 0){
            return null;
        } else{
            return entry.value;
        }
    }
}


//
// API
//

var ht = new HashTable(1024); // size needs to be a power of 2

ht.inset("one", "1");
ht.inset("two", "2");
ht.inset("three", "3");
ht.inset("four", "4");

console.log(ht.get("one")); // '1'
console.log(ht.get("two")); // '2'
console.log(ht.get("three")); // '3'
console.log(ht.get("four")); // '4'

File: Programming_cheat_sheets/_compiled_languages/C/data_structures/hash/hashTable.c
#include <stdlib.h>
#include <stdio.h>
#include <limits.h>
#include <string.h>

//
// Hash Table
//

typedef struct Node{
    char * key;
    char * value;
    struct Node * next;
} Node;


typedef struct hash{
    int size;
    Node ** array; // array of pointers
} Hash;


/**
 * Create a new hash. 
 */
Hash * createHash(int size){

    // Allocate the hash table itself. 
    Hash * hash = malloc(sizeof(Hash));

    // Allocate pointers to the head nodes.
    hash->array = malloc(sizeof(Node *) * size);

    hash->size = size;

    // init the array
    int i;
    for(i = 0; i < size; i++){
        hash->array[i] = NULL;
    }

    return hash;
}

/**
 * Hash a string to get an index for the underlying array `hash->array` 
 */
int hashFunction(Hash * hash, char * key){

    unsigned long int hash_value = 0;
    int i = 0;

    /**
     * Convert our key to a big number.
     * Loop through the letters of the key 
     * while the hash_value is less than ULONG_MAX
     * multiplying hash_value by 256 
     * and adding the int value of the character
     */
    while( i < strlen(key) && hash_value < ULONG_MAX ){
        hash_value *= 256; // same as: `hash_value << 8`
        hash_value += key[i];
        i++;
    }

    return hash_value % hash->size;
}

/**
 * Create an node. 
 */
Node * createNode(char * key, char * value){
    Node * node = (Node *)malloc(sizeof(Node));
    node->key = strdup(key);
    node->value = strdup(value);
    node->next = NULL;

    return node;
}

/**
 * Insert a node into a hash table. 
 */
void insert(Hash * hash, char * key, char * value){

    int index = hashFunction(hash, key);

    Node * currentNode = hash->array[index];

    while(currentNode != NULL && currentNode->key != NULL && strcmp(key, currentNode->key) > 0){
        currentNode = currentNode->next;
    }

    // If there's already a node.  Let's replace that string.
    if(currentNode != NULL && currentNode->key != NULL && strcmp(key, currentNode->key) == 0){
        free(currentNode->value);
        currentNode->value = strdup(value);
    } else { 
        // Nope, could't find it. add a node.
        
        Node * newNode = createNode(key, value);

        // We're at the start of the linked list in this index.
        if(currentNode == hash->array[index]){
            newNode->next = currentNode;
            hash->array[index] = newNode;

        // We're at the end of the linked list in this index.
        } else if (currentNode == NULL){
            currentNode->next = newNode;

        // We're in the middle of the list.
        } else {
            newNode->next = currentNode;
            currentNode->next = newNode;
        }
    }
}

/* Retrieve an node from a hash table. */
char * get(Hash * hash, char * key){

    int index = hashFunction(hash, key);

    Node * currentNode = hash->array[index];

    /* Step through the index, looking for our value. */
    while(currentNode != NULL && currentNode->key != NULL && strcmp(key, currentNode->key) > 0){
        currentNode = currentNode->next;
    }

    // Did we actually find anything?
    if(currentNode == NULL || currentNode->key == NULL || strcmp(key, currentNode->key) != 0){
        return NULL;
    }else{
        return currentNode->value;
    }
}

int main(){
    Hash * hash = createHash(100); 

    insert(hash, "name", "brian");
    insert(hash, "address", "123 4th Ave");
    insert(hash, "country", "USA");
    insert(hash, "state", "AZ");

    printf("%s\n", get(hash, "name")); // brian
    printf("%s\n", get(hash, "address")); // 123 4th Ave
    printf("%s\n", get(hash, "country")); // USA
    printf("%s\n", get(hash, "state")); // AZ
    
    return 0;
}
